pkg/vrf/grpc.go

```diff
@@ -1,218 +1,196 @@
 		return nil, err
 	}
 	return response, nil
 }
 
 // DeleteVrf deletes a VRF
-func (s *Server) DeleteVrf(ctx context.Context, in *pb.DeleteVrfRequest) (*emptypb.Empty, error) {
+func (s *Server) DeleteVrf(_ context.Context, in *pb.DeleteVrfRequest) (*emptypb.Empty, error) {
 	// check input correctness
 	if err := s.validateDeleteVrfRequest(in); err != nil {
+		log.Printf("DeleteVrf(): validation failure: %v", err)
 		return nil, err
 	}
 	// fetch object from the database
-	obj := new(pb.Vrf)
-	ok, err := s.store.Get(in.Name, obj)
+	_, err := s.getVrf(in.Name)
 	if err != nil {
-		fmt.Printf("Failed to interact with store: %v", err)
-		return nil, err
-	}
-	if !ok {
-		if in.AllowMissing {
-			return &emptypb.Empty{}, nil
+		if err != infradb.ErrKeyNotFound {
+			log.Printf("Failed to interact with store: %v", err)
+			return nil, err
 		}
-		err := status.Errorf(codes.NotFound, "unable to find key %s", in.Name)
-		return nil, err
-	}
-	// configure netlink
-	if err := s.netlinkDeleteVrf(ctx, obj); err != nil {
-		return nil, err
-	}
-	// delete from FRR
-	if err := s.frrDeleteVrfRequest(ctx, obj); err != nil {
-		return nil, err
+		if !in.AllowMissing {
+			err = status.Errorf(codes.NotFound, "unable to find key %s", in.Name)
+			log.Printf("DeleteVrf(): Vrf with id %v: Not Found %v", in.Name, err)
+			return nil, err
+		}
+		return &emptypb.Empty{}, nil
 	}
-	// remove from the Database
-	delete(s.ListHelper, obj.Name)
-	err = s.store.Delete(obj.Name)
-	if err != nil {
+
+	if err := s.deleteVrf(in.Name); err != nil {
+		log.Printf("DeleteVrf(): Vrf with id %v, Delete Vrf from DB failure: %v", in.Name, err)
 		return nil, err
 	}
+
 	return &emptypb.Empty{}, nil
 }
 
 // UpdateVrf updates an VRF
-func (s *Server) UpdateVrf(ctx context.Context, in *pb.UpdateVrfRequest) (*pb.Vrf, error) {
+func (s *Server) UpdateVrf(_ context.Context, in *pb.UpdateVrfRequest) (*pb.Vrf, error) {
 	// check input correctness
 	if err := s.validateUpdateVrfRequest(in); err != nil {
+		log.Printf("UpdateVrf(): validation failure: %v", err)
 		return nil, err
 	}
 	// fetch object from the database
-	obj := new(pb.Vrf)
-	ok, err := s.store.Get(in.Vrf.Name, obj)
+	vrfObj, err := s.getVrf(in.Vrf.Name)
 	if err != nil {
-		fmt.Printf("Failed to interact with store: %v", err)
-		return nil, err
-	}
-	if !ok {
-		// TODO: introduce "in.AllowMissing" field. In case "true", create a new resource, don't return error
-		err := status.Errorf(codes.NotFound, "unable to find key %s", in.Vrf.Name)
-		return nil, err
+		if err != infradb.ErrKeyNotFound {
+			log.Printf("UpdateVrf(): Failed to interact with store: %v", err)
+			return nil, err
+		}
+		if !in.AllowMissing {
+			err = status.Errorf(codes.NotFound, "unable to find key %s", in.Vrf.Name)
+			log.Printf("UpdateVrf(): Vrf with id %v: Not Found %v", in.Vrf.Name, err)
+			return nil, err
+		}
+
+		log.Printf("UpdateVrf(): Vrf with id %v is not found so it will be created", in.Vrf.Name)
```

Comment:
Should the Update really do an auto-create if the port is not found or should it just return not found? (see port/grpc.go comment)

Reply:
Answered on that above